storage/filestore/dir.go

```diff
@@ -273,48 +265,61 @@ func (dir *Dir) Delete(ctx context.Context, ref storage.BlobRef) (err error) {
 	// _forwards_, this race should not occur because it is assumed that pieces are never
 	// rewritten with an _older_ storage format version.
 	for i := MinFormatVersionSupported; i <= MaxFormatVersionSupported; i++ {
-		verPath := blobPathForFormatVersion(pathBase, i)
-
-		// move to trash folder, this is allowed for some OS-es
-		moveErr = rename(verPath, trashPath)
-		if os.IsNotExist(moveErr) {
-			// no piece at that path; either it has a different storage format version or there
-			// was a concurrent delete. (this function is expected by callers to return a nil
-			// error in the case of concurrent deletes.)
-			continue
-		}
-		if moveErr != nil {
-			// piece could not be moved into the trash dir; we'll try removing it directly
-			trashPath = verPath
-		}
+		combinedErrors.Add(dir.DeleteWithStorageFormat(ctx, ref, i))
+	}
 
-		// try removing the file
-		err = os.Remove(trashPath)
+	return combinedErrors.Err()
+}
 
-		// ignore concurrent deletes
-		if os.IsNotExist(err) {
-			// something is happening at the same time as this; possibly a concurrent delete,
-			// or possibly a rewrite of the blob. keep checking for more versions.
-			continue
-		}
+// DeleteWithStorageFormat deletes the blob with the specified ref for one specific format version
+func (dir *Dir) DeleteWithStorageFormat(ctx context.Context, ref storage.BlobRef, formatVer storage.FormatVersion) (err error) {
+	defer mon.Task()(&ctx)(&err)
 
-		// the remove may have failed because of an open file handle. put it in a queue to be
-		// retried later.
-		if err != nil {
-			dir.mu.Lock()
-			dir.deleteQueue = append(dir.deleteQueue, trashPath)
-			dir.mu.Unlock()
-		}
+	pathBase, err := dir.blobToBasePath(ref)
+	if err != nil {
+		return err
+	}
+	trashPath := dir.blobToTrashPath(ref)
 
-		// ignore is-busy errors, they are still in the queue
-		// but no need to notify
-		if isBusy(err) {
-			err = nil
-		}
-		combinedErrors.Add(err)
+	verPath := blobPathForFormatVersion(pathBase, formatVer)
+
+	// move to trash folder, this is allowed for some OS-es
+	moveErr := rename(verPath, trashPath)
+	if os.IsNotExist(moveErr) {
```

Comment:
It's just occurred to me that this is not a great check- it might also mean that some path component prefix of `trashPath` does not exist, instead of saying `verPath` could not be found. If someone mistakenly deletes the trash dir, we don't want all file deletions to start becoming no-ops! (Yes, this problem was present in the prior code too. Good time to fix it though.)

Can we at least do a stat on `dir.garbagedir()` at the top of `Delete()`? Just as a sort of sanity check, because ensuring that the dir still exists by the time we get to the `rename()` step would be near impossible.

Reply:
That makes sense. I would say that we should do the check in `blobToTrashPath` (or something like that). That function could return an error if the trash path does not exist, essentially saying that we cannot convert the blob to trash path if the path does not exist.

Either way, we'd need to check each time we call `DeleteWithStorageFormat` rather than `Delete`, because the former can be called directly.

Thoughts? (I'll proceed with my idea, but can change it.)

Reply:
Fixed.